packages/core/services/oidc-provider/oidc-provider.js

```diff
@@ -0,0 +1,133 @@
+
+    this.client = new Redis(this.settings.redisUrl, { keyPrefix: 'oidc:' });
+
+    const { privateJwk } = await this.broker.call('jwk.get');
+
+    const config = baseConfig(this.settings, privateJwk);
+
+    config.adapter = name => new RedisAdapter(name, this.client);
+
+    // See https://github.com/panva/node-oidc-provider/blob/main/recipes/client_based_origins.md
+    config.clientBasedCORS = (ctx, origin, client) => {
+      // TODO validate CORS based on client
+      console.log('origin, client', origin, client);
+      return true;
+    };
+
+    this.oidc = new Provider(this.settings.baseUrl, config);
+
+    // Allow provider to interpret reverse proxy headers.
+    this.oidc.proxy = true;
+
+    await this.broker.call('api.addRoute', {
+      route: {
+        path: '/.well-known/openid-configuration',
+        aliases: {
+          'GET /': 'oidc-provider.proxyConfig'
+        }
+      }
+    });
+
+    await this.broker.call('api.addRoute', {
+      route: {
+        path: '/.oidc/auth',
+        use: [this.oidc.callback()]
+      }
+    });
+  },
+  actions: {
+    // See https://moleculer.services/docs/0.13/moleculer-web.html#Authentication
+    async authenticate(ctx) {
+      const { route, req, res } = ctx.params;
+      // Extract token from authorization header (do not take the Bearer part)
+      const token = req.headers.authorization && req.headers.authorization.split(' ')[1];
+      if (token) {
+        const payload = await ctx.call('jwk.verifyToken', { token });
+        if (payload) {
+          ctx.meta.tokenPayload = payload;
+          ctx.meta.webId = payload.webid; // Not webId !!
+          return Promise.resolve(payload);
+        }
+        // Invalid token
+        // TODO make sure token is deleted client-side
+        ctx.meta.webId = 'anon';
+        return Promise.reject(new E.UnAuthorizedError(E.ERR_INVALID_TOKEN));
+      }
+      // No token
+      ctx.meta.webId = 'anon';
+      return Promise.resolve(null);
+    },
+    // See https://moleculer.services/docs/0.13/moleculer-web.html#Authorization
+    async authorize(ctx) {
+      const { route, req, res } = ctx.params;
+      // Extract token from authorization header (do not take the Bearer part)
+      const token = req.headers.authorization && req.headers.authorization.split(' ')[1];
+      if (token) {
+        const payload = await ctx.call('jwk.verifyToken', { token });
+        if (payload) {
+          ctx.meta.tokenPayload = payload;
+          ctx.meta.webId = payload.webid; // Not webId !!
+          return Promise.resolve(payload);
+        }
+        ctx.meta.webId = 'anon';
+        return Promise.reject(new E.UnAuthorizedError(E.ERR_INVALID_TOKEN));
+      }
+      ctx.meta.webId = 'anon';
+      return Promise.reject(new E.UnAuthorizedError(E.ERR_NO_TOKEN));
+    },
+    async proxyConfig() {
+      const res = await fetch(`${this.settings.baseUrl}.oidc/auth/.well-known/openid-configuration`);
```

Comment:
Why the fetch?

Reply:
The node-oidc-provider library use koa internally, and I haven't found a way to proxy directly to this route.